src/NetworkStatus.tsx

```diff
@@ -0,0 +1,81 @@
+import React, {
+  createContext,
+  PropsWithChildren,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { Network } from "@capacitor/network";
+import { PluginListenerHandle } from "@capacitor/core";
+import { onlineManager } from "@tanstack/react-query";
+
+interface NetworkStatusContextValue {
+  isOnline: boolean;
+}
+
+const NetworkStatusContext = createContext<NetworkStatusContextValue>({
+  isOnline: true,
+});
+
+const NetworkStatusProvider: React.FC<PropsWithChildren> = ({ children }) => {
+  const [isOnline, setIsOnline] = useState<boolean | null>(null);
+
+  useEffect(() => {
+    // Used to prevent setting up the listener from being added from an unmounted component
```

Comment:
It is to avoid having multiple listeners be attached simultaneously. The teardown function returned by the effect will remove the listener _if_ it has been created by the time the teardown function is called. But because the setup function is async, it's possible for the component to be mounted and unmounted all while the async action is still happening.